lib/loglimit/loglimit.go

```diff
@@ -0,0 +1,205 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package loglimit
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	// timeWindow is the time window over which logs are deduplicated.
+	timeWindow = time.Minute
+	// timeWindowCleanupInterval is the interval between cleanups of time
+	// windows that have already ended.
+	// Since the time window is set to 1 minute and the cleanup interval
+	// is set to 10 seconds, time windows can be in fact slightly larger
+	// than 1 minute (i.e., 1 minute and 10 seconds, in the worst case).
+	timeWindowCleanupInterval = 10 * time.Second
+)
+
+// Config contains the log limiter config.
+type Config struct {
+	// Context is a context to signal stops, cancellations.
+	Context context.Context
+	// MessageSubstrings contains a list of substrings belonging to the
+	// logs that should be deduplicated.
+	MessageSubstrings []string
+	// Clock is a clock to override in tests, set to real time clock
+	// by default.
+	Clock clockwork.Clock
+}
+
+// checkAndSetDefaults verifies configuration and sets defaults
+func (c *Config) checkAndSetDefaults() error {
+	if c.Context == nil {
+		c.Context = context.TODO()
```

Comment:
Don't use `context.TODO()` unless there is a follow-up to be done.

```suggestion
		c.Context = context.Background()
```